xonsh/procs/executables.py

```diff
@@ -0,0 +1,112 @@
+"""Interfaces to locate executable files on file system."""
+
+import os
+from pathlib import Path
+
+from xonsh.built_ins import XSH
+from xonsh.platform import ON_WINDOWS
+
+
+def get_possible_names(name, env=None):
+    """Expand name to all possible variants based on `PATHEXT`.
+
+    PATHEXT is a Windows convention containing extensions to be
+    considered when searching for an executable file.
+
+    Conserves order of any extensions found and gives precedence
+    to the bare name.
+    """
+    env = env if env is not None else XSH.env
+    env_pathext = env.get("PATHEXT", [])
+    if not env_pathext:
+        return [name]
+    upper = name.upper() == name
+    extensions = [""] + env_pathext
+    return [name + (ext.upper() if upper else ext.lower()) for ext in extensions]
+
+
+def clear_paths(paths):
+    """Remove duplicates and nonexistent directories from paths."""
+    cont = set()
+    for p in map(os.path.realpath, paths):
+        if p not in cont:
+            cont.add(p)
+            if os.path.isdir(p):
+                yield p
```

Comment:
This is the `unique_everseen` pattern:

```suggestion
    return filter(os.path.isdir, unique_everseen(map(os.path.realpath, paths)))
```

Unique_everseen is available in [more_itertools](https://pypi.org/project/more-itertools) or copy/pasted from stdlib docs into xonsh.lib.itertools.